biogeochem/FatesAllometryMod.F90

```diff
@@ -2625,7 +2949,265 @@ subroutine cspline(x1,x2,y1,y2,dydx1,dydx2,x,y,dydx)
+         if (converged) exit regfalsi_loop
+         !---~---
+
+
+         !--- Find the new function evaluation.
+         call h_allom(rdbh,ipft,hgt)
+         rfun = rdbh * rdbh * hgt - size
+         !---~---
+
+
+         !---~---
+         !   Define the new searching interval based on the intermediate value theorem.
+         !---~---
+         if (abs(rfun) < nearzero) then
+            !--- Converged by luck.
+            converged = .true.
+            exit regfalsi_loop
+            !---~---
+         else if (rfun * afun <= -nearzero ) then
+            !--- Guess is between lower and current guess.
+            zdbh = rdbh
+            zfun = rfun
+            !--- If we are updating the upper side again, halve afun (Regula Falsi method).
+            if (zside) afun = afun * 0.5_r8
+            !--- Flag that we have just updated the upper side.
+            zside = .true.
+            !---~---
+         else
+            !--- Guess is between current and upper guess.
+            adbh = rdbh
+            afun = rfun
+            !--- If we are updating the lower side again, halve zfun (Regula Falsi method).
+            if (.not. zside) zfun = zfun * 0.5_r8
+            !--- Flag that we have just updated the lower side.
+            zside = .false.
+         end if
+      end do regfalsi_loop
+      !---~---
+
+
+      !---~---
+      !   Check that the Regula Falsi method indeed converged.
+      !---~---
+      if (converged) then
+         !--- Yes, return the last guess
+         dbh = rdbh
+         !---~---
+      else
+         !--- No, report the bad news
+         write (unit=*,fmt='(a)')           '---~---'
+         write (unit=*,fmt='(a)')           ' Failed finding the solution:'
+         write (unit=*,fmt='(a)')           '---~---'
+         write (unit=*,fmt='(a)')           ' '
+         write (unit=*,fmt='(a)')           ' Input:   '
+         write (unit=*,fmt='(a,1x,es14.7)') ' + size  =',size
+         write (unit=*,fmt='(a,1x,es14.7)') ' + dbh   =',dbh
+         write (unit=*,fmt='(a)')           ' '
+         write (unit=*,fmt='(a)')           ' Current guesses and evaluations:'
+         write (unit=*,fmt='(a,1x,es14.7)') ' + adbh  =',adbh
+         write (unit=*,fmt='(a,1x,es14.7)') ' + afun  =',afun
+         write (unit=*,fmt='(a,1x,es14.7)') ' + rdbh  =',rdbh
+         write (unit=*,fmt='(a,1x,es14.7)') ' + rfun  =',rfun
+         write (unit=*,fmt='(a,1x,es14.7)') ' + zdbh  =',zdbh
+         write (unit=*,fmt='(a,1x,es14.7)') ' + zfun  =',zfun
+         write (unit=*,fmt='(a)')           ' '
+         write (unit=*,fmt='(a)')           '---~---'
+         write(fates_log(),*) 'Size to DBH routine failed to converge!'
+         call endrun(msg=errMsg(sourcefile, __LINE__))
+         !---~---
+      end if
+      !---~---
+
+      return
+   end subroutine size2dbh
+   ! ============================================================================
+
```

Comment:
You were very determined to find the inverse solution for dbh and h @mpaiao , not easy, kudos

Reply:
This one is totally borrowed from ED2 solvers, I just adapted it here. This allometric function is well behaved and should quickly converge with the Newton method, but I thought I should keep the full root finding approach in case we want to adapt for less well-behaved functions anywhere in FATES in the future.